gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -451,7 +459,21 @@ private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, SourceSt
         LOG.warn(
             offsetNotFoundMsg + "This partition will start from the earliest offset: " + offsets.getEarliestOffset());
         offsets.startAtEarliestOffset();
-      } else {
+      } else if (offsetOption.equals(OFFSET_LOOKBACK)) {
+        long lookbackOffsetRange = state.getPropAsLong("kafka.offset.lookback", 0L);
+        long offset = offsets.getLatestOffset() - lookbackOffsetRange;
+        LOG.warn(offsetNotFoundMsg + "This partition will start from latest-lookback [ " + offsets.getLatestOffset() + " - " + lookbackOffsetRange + " ]  start offset: " + offset);
+        try {
+          offsets.startAt(offset);
+        } catch (StartOffsetOutOfRangeException e) {
+          String offsetOutOfRangeMsg = String.format(
+                  "Start offset for partition %s is out of range. Start offset = %d, earliest offset = %d, latest offset = %d.",
+                  partition, offsets.getStartOffset(), offsets.getEarliestOffset(), offsets.getLatestOffset());
+          LOG.warn(offsetOutOfRangeMsg + "This partition will start from the latest offset: " + offsets.getLatestOffset());
```

Comment:
Can have race condition (latest offset can change) between this log message and the next startAtLatestOffsets call. 

So get latest offset first. 
Then log it and start consuming from it as a provided offset?

Reply:
I have made changes as per the below comment from autumnust. I am reusing the functionality to check for "RESET_ON_OFFSET_OUT_OF_RANGE" and make decision accordingly

Reply:
We are using this functionality (with lookback) in our production pipeline since many days. It is going to be very useful where data retention is couple of days in kafka and wants to avoid using earliest. Lookback help us in going back to certain offset and start consuming from that point (when we lost previous state information).